File: golang/21_merge_two_sorted_lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	h3 := &ListNode{
		Next: nil,
		Val:  -999999,
	}
	t3 := &ListNode{
		Next: nil,
		Val:  -999999,
	}
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			temp := &ListNode{
				Next: nil,
				Val:  l1.Val,
			}
			if t3.Val != -999999 {
				t3.Next = temp
				t3 = temp
			} else {
				h3 = temp
				t3 = temp
			}
			l1 = l1.Next
		} else {
			temp := &ListNode{
				Next: nil,
				Val:  l2.Val,
			}
			if t3.Val != -999999 {
				t3.Next = temp
				t3 = temp
			} else {
				h3 = temp
				t3 = temp
			}
			l2 = l2.Next
		}
	}
	for l1 != nil {
		temp := &ListNode{
			Next: nil,
			Val:  l1.Val,
		}
		if t3.Val != -999999 {
			t3.Next = temp
			t3 = temp
		} else {
			h3 = temp
			t3 = temp
		}
		l1 = l1.Next
	}
	for l2 != nil {
		temp := &ListNode{
			Next: nil,
			Val:  l2.Val,
		}
		if t3.Val != -999999 {
			t3.Next = temp
			t3 = temp
		} else {
			h3 = temp
			t3 = temp
		}
		l2 = l2.Next
	}
	if h3.Val == -999999 {
		return nil
	}
	return h3
}

File: python/21_merge_two_sorted_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        head1 = l1
        head2 = l2
        head_merged, tail_merged = None, None
        while head1 != None and head2 != None:
            if head1.val < head2.val:
                temp = ListNode(head1.val)
                if tail_merged != None:
                    tail_merged.next = temp
                    tail_merged = temp
                else:
                    head_merged = temp
                    tail_merged = temp
                head1 = head1.next
            else:
                temp = ListNode(head2.val)
                if tail_merged != None:
                    tail_merged.next = temp
                    tail_merged = temp
                else:
                    head_merged = temp
                    tail_merged = temp
                head2 = head2.next
        while head1:
            temp = ListNode(head1.val)
            if tail_merged != None:
                tail_merged.next = temp
                tail_merged = temp
            else:
                head_merged = temp
                tail_merged = temp
            head1 = head1.next
        while head2:
            temp = ListNode(head2.val)
            if tail_merged != None:
                tail_merged.next = temp
                tail_merged = temp
            else:
                head_merged = temp
                tail_merged = temp
            head2 = head2.next
        return head_merged